### ISSUE TITLE
Hwmon instrument doesn't pick up sensors with new driver

### ISSUE THREAD

**Issue by credp** [ARM-software/workload-automation#150]
In the latest LSKs (and mainline), the names of the sensors have changed. They no longer contain the cpu types, the full list is below.

I have been able to work around this in lisa by specifying the name of the sensors I wish to track, but I cannot do this in WA.

The sensors and names available are:
curr1_label - "BOARD_CURR_SYS"
curr2_label - "BOARD_CURR_BIG"
curr3_label - "BOARD_CURR_LITTLE"
curr4_label - "BOARD_CURR_GPU"
energy1_label - "BOARD_ENERGY_SYS"
energy2_label - "BOARD_ENERGY_BIG"
energy3_label - "BOARD_ENERGY_LITTLE"
energy4_label - "BOARD_ENERGY_GPU"
in0_label - "SENSOR_VOLT_BIG"
in1_label - "SENSOR_VOLT_LITTLE"
in2_label - "SENSOR_VOLT_SYS"
in3_label - "BOARD_VOLT_SYS"
in4_label - "BOARD_VOLT_BIG"
in5_label - "BOARD_VOLT_LITTLE"
in6_label - "BOARD_VOLT_GPU"
power1_label - "BOARD_POWER_SYS"
power2_label - "BOARD_POWER_BIG"
power3_label - "BOARD_POWER_LITTLE"
power4_label - "BOARD_POWER_GPU"
temp1_label - "SENSOR_TEMP_PMIC"
temp2_label - "SENSOR_TEMP_SOC"
temp3_label - "SENSOR_TEMP_BIG"
temp4_label - "SENSOR_TEMP_LITTLE"
temp5_label - "SENSOR_TEMP_GPU_0"
temp6_label - "SENSOR_TEMP_GPU_1"

Thanks :)

**Comment by credp** [comment]
[0002-HwmonDiscovery-Change-logic-for-handling-non-device-.patch.tar.gz](https://github.com/ARM-software/workload-automation/files/240483/0002-HwmonDiscovery-Change-logic-for-handling-non-device-.patch.tar.gz)

Here's my local solution as a patch. I am wary of proposing a pull request simply because I had to change the name gathering behaviour, but this code used to work so I am a little unsure what's going on.

I also think at least one phase of a better solution would be to try to assert that a file actually exists
and raise an exception directly from get_sysfile_value, although that would require much more testing.
The code here looks as though it assumes that would be mapped to a DeviceError, although it doesn't
happen with a Juno device.

**Comment by ep1cman** [comment]
Which linaro version are you using? I can't seem to get the latest (16.04) to work on my juno

**Comment by ep1cman** [comment]
When I tried to run this get_sysfile_value does correctly raise a DeviceError, there is no need to check the files existence and raise a separate error:

ipdb> device_path = path.join(HWMON_ROOT, hwmon_device, 'device')
ipdb> device_path
'/sys/class/hwmon/hwmon0/device'
ipdb>  name = device.get_sysfile_value(path.join(device_path, 'name'))
2016-05-04 16:12:25,733 DEBUG    ssh: cat '/sys/class/hwmon/hwmon0/device/name'
*** DeviceError: Got exit code 1
from: cat '/sys/class/hwmon/hwmon0/device/name'
OUTPUT: cat: can't open '/sys/class/hwmon/hwmon0/device/name': No such file or directory

**Comment by ep1cman** [comment]
Could you elaborate on what the issue is here? It seem deal with the new sensor names fine for me,

The reason we are appending the device name onto the sensor label is that on other devices they have hwmon devices per sensor and we need a way to differentiate between them,

**Comment by credp** [comment]
So, the version I had originally (6c4d88ff) did not detect any sensors for hwmon at all. The output in the log while it was looking was like:

```
2016-04-26 14:17:58,499 DEBUG    HwmonInstrument: Searching for HWMON sensors.
2016-04-26 14:17:58,499 DEBUG    android: adb -s 10.1.207.30:5555 shell "ls /sys/class/hwmon"
2016-04-26 14:17:58,526 DEBUG    android: adb -s 10.1.207.30:5555 shell "echo 'cat '\\''/sys/class/hwmon/hwmon0/device/name'\\''' | su"
2016-04-26 14:17:58,561 DEBUG    android: adb -s 10.1.207.30:5555 shell "if [ -e '/sys/class/hwmon/hwmon0/device/energy1_input' ]; then echo 1; else echo 0; fi"
2016-04-26 14:17:58,578 DEBUG    android: adb -s 10.1.207.30:5555 shell "if [ -e '/sys/class/hwmon/hwmon0/device/temp1_input' ]; then echo 1; else echo 0; fi"
2016-04-26 14:17:58,593 DEBUG    android: adb -s 10.1.207.30:5555 shell "echo 'cat '\\''/sys/class/hwmon/hwmon1/device/name'\\''' | su"
2016-04-26 14:17:58,621 DEBUG    android: adb -s 10.1.207.30:5555 shell "if [ -e '/sys/class/hwmon/hwmon1/device/energy1_input' ]; then echo 1; else echo 0; fi"
2016-04-26 14:17:58,642 DEBUG    android: adb -s 10.1.207.30:5555 shell "if [ -e '/sys/class/hwmon/hwmon1/device/temp1_input' ]; then echo 1; else echo 0; fi"
```

The output files did not monitor any sensors. I looked into it, and the only reason I could find was that there was no DeviceError raised, so we never dropped into the hwmon0 folder rather than hwmon0/device.

I added this myself, as you saw in the patch (and it's madness to do it there, I agree, but I needed numbers asap), and then I saw that the sensor names were never reset and I ended up with this set of names:

```
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS	119.439694	Joules
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG	6.788674	Joules
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE	14.210176	Joules
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU	19.784242	Joules
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC	42.85	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC before	42.85	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC after	42.85	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC	34.8785	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC before	34.843	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC after	34.914	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC SENSOR_TEMP_BIG	35.609	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC SENSOR_TEMP_BIG before	35.07	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC SENSOR_TEMP_BIG after	36.148	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC SENSOR_TEMP_BIG SENSOR_TEMP_LITTLE	36.17	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC SENSOR_TEMP_BIG SENSOR_TEMP_LITTLE before	36.143	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC SENSOR_TEMP_BIG SENSOR_TEMP_LITTLE after	36.197	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC SENSOR_TEMP_BIG SENSOR_TEMP_LITTLE SENSOR_TEMP_GPU_0	34.8505	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC SENSOR_TEMP_BIG SENSOR_TEMP_LITTLE SENSOR_TEMP_GPU_0 before	34.638	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC SENSOR_TEMP_BIG SENSOR_TEMP_LITTLE SENSOR_TEMP_GPU_0 after	35.063	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC SENSOR_TEMP_BIG SENSOR_TEMP_LITTLE SENSOR_TEMP_GPU_0 SENSOR_TEMP_GPU_1	34.9495	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC SENSOR_TEMP_BIG SENSOR_TEMP_LITTLE SENSOR_TEMP_GPU_0 SENSOR_TEMP_GPU_1 before	34.749	Celsius
mp_eas_interactive_w01	bbench_with_audio	1	scpi_sensors BOARD_ENERGY_SYS BOARD_ENERGY_BIG BOARD_ENERGY_LITTLE BOARD_ENERGY_GPU SENSOR_TEMP_PMIC SENSOR_TEMP_SOC SENSOR_TEMP_BIG SENSOR_TEMP_LITTLE SENSOR_TEMP_GPU_0 SENSOR_TEMP_GPU_1 after	35.15	Celsius

```

Hence the addition of clearing the name.

In regard to the DeviceError, I have two implementations of get_sysfile_value in my codebase - one in common/linux/device.py which is implemented like so:
```

    def get_sysfile_value(self, sysfile, kind=None):
        """
        Get the contents of the specified sysfile.

        :param sysfile: The file who's contents will be returned.

        :param kind: The type of value to be expected in the sysfile. This can
                     be any Python callable that takes a single str argument.
                     If not specified or is None, the contents will be returned
                     as a string.

        """
        output = self.execute('cat \'{}\''.format(sysfile), as_root=self.is_rooted).strip()  # pylint: disable=E1103
        if kind:
            return kind(output)
        else:
            return output

```

And another is in core/device.py, which is just the template.

```
    def get_sysfile_value(self, sysfile, kind=None):
        """
        Get the contents of the specified sysfile.

        :param sysfile: The file who's contents will be returned.

        :param kind: The type of value to be expected in the sysfile. This can
                     be any Python callable that takes a single str argument.
                     If not specified or is None, the contents will be returned
                     as a string.

        """
        raise NotImplementedError()

```

Maybe there is something in self.execute which is different?

**Comment by ep1cman** [comment]
What was the Juno setup you had? This is indeed very messed up, if I can recreate your Juno setup I could look into it properly because my current setup doesn't have this issue

**Comment by credp** [comment]
I've got Linaro Android HMP (and EAS) kernels from 16.02 plus Linaro Android userspace at that release. I had some problems building later ones so I stuck there. I'm using an r2 Juno board as well, which might be significant. I'm just downstairs if you need me.

**Comment by ep1cman** [comment]
This should be resolved by #157, can you give it a try and let us know if it now works correctly for you?

**Comment by credp** [comment]
I had a little time spare, so I tried it out now. It seems to be working fine except that having a space in the sensor name is a bit awkward for me. I can probably deal with it though :) I already have to have logic to deal with mapping one name to another so I can compare results from older kernels with new ones.

scpi_sensors BOARD_ENERGY_SYS = 73.8984 Joules
scpi_sensors BOARD_ENERGY_BIG = 74.091557 Joules
scpi_sensors BOARD_ENERGY_LITTLE = 24.97484 Joules
scpi_sensors BOARD_ENERGY_GPU = 11.245183 Joules

**Comment by ep1cman** [comment]
After a chat with @setrofim we think the way the hwmon metrics are reported should be changed a bit. Our thoughts are:
 - hwmon devices should be added a classifier to the metrics
 - the name of the metric should be the label of the sensor, but if there isn't a label it would be composed of the sensor type & id e.g. temp4

What are your thoughts/suggestions on this?

**Comment by credp** [comment]
I agree. In most cases I think the label is the important part and it is the part you are likely to filter for when analysing results.
I'm having difficulty understanding exactly what you mean though - so for example, would an hwmon line in results.csv

**Comment by ep1cman** [comment]
An example of the new results.csv output would be:
![image](https://cloud.githubusercontent.com/assets/5303409/15107413/989ed484-15c7-11e6-9656-1a5fa057c8c7.png)

**Comment by credp** [comment]
That looks great to me, I can work very easily with that :) Thanks!

**Comment by ep1cman** [comment]
Resolved by 9296bafbd98d1d1d050987fdcdd193f8e0b8583b
